File: generate-parentheses/generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList<>();
        backtrack(result, n, "", 0, 0);
        return result;
    }
    
    void backtrack(List<String> result, int n, String temp, int open, int close) {
        if (temp.length() == n*2) {
            result.add(temp);
        }
        if (open < n) {
            backtrack(result, n, temp + "(", open + 1, close);
        }
        if (close < open) {
            backtrack(result, n, temp + ")", open, close + 1);
        }
    }
}

File: generate-parentheses/generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        result = []
        self.backtrack(n, 0, 0, result, "")
        return result
    def backtrack(self, n, open_int, close, result, temp):
        if close == n:
            result.append(temp)
        if open_int < n:
            self.backtrack(n, open_int + 1, close, result, temp + "(")
        if close < open_int:
            self.backtrack(n, open_int, close + 1, result, temp + ")")
            